docs/advanced.rst

```diff
@@ -0,0 +1,523 @@
+            actors.
+            """
+
+
+To "duck type" this class, you must implement the following methods:
+
+* ``on_after_load_track``
+    This method is called by Rally after a track has been loaded by Rally but before work is distributed to worker actors. Implementations are expected to modify the
+    provided track object in place.
+* ``on_prepare_track``
+    This method is called by Rally after the ``on_after_load_track`` phase. Here, any data that is necessary for
+    benchmark execution should be prepared, e.g. by downloading data or generating it. Implementations should
+    be aware that this method might be called on a different machine than ``on_after_load_track`` and they cannot
+    share any state in between phases. The method should `yield` a tuple of a Callable and its parameters for each
+    Track Processor thread to be able to work in parallel.
+
+Consider the DefaultTrackPreparator below, which is invoked by default unless overridden by custom registered track processors::
+
+    class DefaultTrackPreparator(TrackProcessor):
+        def __init__(self):
+            super().__init__()
+            # just declare here, will be injected later
+            self.cfg = None
+            self.downloader = None
+            self.decompressor = None
+            self.track = None
+
+        def on_after_load_track(self, track):
+            ...
+
+        @staticmethod
+        def prepare_docs(cfg, track, corpus, preparator):
+            for document_set in corpus.documents:
+                if document_set.is_bulk:
+                    data_root = data_dir(cfg, track.name, corpus.name)
+                    logging.getLogger(__name__).info(
+                        "Resolved data root directory for document corpus [%s] in track [%s] to [%s].", corpus.name, track.name, data_root
+                    )
+                    if len(data_root) == 1:
+                        preparator.prepare_document_set(document_set, data_root[0])
+                    # attempt to prepare everything in the current directory and fallback to the corpus directory
+                    elif not preparator.prepare_bundled_document_set(document_set, data_root[0]):
+                        preparator.prepare_document_set(document_set, data_root[1])
+
+        def on_prepare_track(self, track, data_root_dir):
+            prep = DocumentSetPreparator(track.name, self.downloader, self.decompressor)
+            for corpus in used_corpora(track):
+                params = {"cfg": self.cfg, "track": track, "corpus": corpus, "preparator": prep}
+                yield DefaultTrackPreparator.prepare_docs, params
+
+In this case, you can see by default we do nothing here for ``on_after_load_track`` to mutate the track, but yield a tuple of the ``prepare_docs``
+function and its parameters for each corpus in the track ``corpora``. After this is called, these tuples are given to each TrackProcessor worker actor
+to be executed in parallel.
```

Comment:
Should we explain a bit more what the expected params typically are? e.g. here we see the config object, the track, which are easy to understand, but it's unclear what the `DocumentSetPreparator` does (passed via `preparator`). A _high level_ explanation would be beneficial in my opinion.